### PR TITLE
gcp_iit: add instance metadata and label selectors

### LINKED ISSUE #1001: Expand GCP selectors to include labels and metadata
GCP node attestation currently produces a selector per instance tag but does not include selectors for labels and metadata. Selecting on certain labels and metadata may make sense in some deployment environments.

The GCP node attestor should be updated to allow specific labels/metadata keys to contribute to the selectors. It probably isn't a good idea to just add selectors for all labels and metadata since the list of items and their values can be arbitrarily large (metadata values themselves can be as large as 256KiB).

I'd propose we add something like the following to the plugin configuration to allow operators to opt in to the metadata and label keys they want to use for node selection: 

```
use_instance_metadata_keys = ["some-metadata-key"]
use_instance_labels = ["some-label-key"]
```

The attestor could then produce selectors for metadata and label keys that match those sets:
```
gcp_iit:instance-metadata:some-metadata-key:some-metadata-value
gcp_iit:instance-label:some-label-key:some-label-value
```

We might consider an upper bound on the size of values for candidate metadata/labels to prevent stuffing large values into the data store.

### PR DESCRIPTION
**Pull request by azdagron**
This PR enhances the GCP IIT node attestor to gather instance metadata and labels into selectors.

The plugin uses an opt-in strategy for metadata keys it will consider since metadata values can be quite large and not all metadata necessarily useful for node selection. Also, a configurable upper bound on metadata values is also enforced. Metadata that exceeds the limit is ignored.

All labels are added as selectors. 

Fixes #1001

### CODE REVIEW

**Comment by azdagron** [comment]
I didn't implement the opt-in approach for labels. If instances can typically have a large label set, maybe it would be a good idea. Thoughts anybody?

**Comment by APTy** [comment]
> I didn't implement the opt-in approach for labels. If instances can typically have a large label set, maybe it would be a good idea. Thoughts anybody?

I think for **Workload Attestation** it's fine to return a large set of selectors, since they're really only used by the Agent for the purpose of identifying a process.

In the case of **Node Attestation**, we have to be careful because we create an extra database entry for each selector. Because of this, I think it's okay to be conservative and ask users to opt-in. At least until someone comes along and says "I want them all", at which point, I'd still at least want a flag to be able to opt-out.

File: pkg/server/plugin/nodeattestor/gcp/iit.go [hunk 1]
```diff
@@ -20,9 +22,10 @@ import (
 )
 
 const (
-	pluginName    = "gcp_iit"
-	tokenAudience = "spire-gcp-node-attestor"
-	googleCertURL = "https://www.googleapis.com/oauth2/v1/certs"
+	pluginName                  = "gcp_iit"
+	tokenAudience               = "spire-gcp-node-attestor"
+	googleCertURL               = "https://www.googleapis.com/oauth2/v1/certs"
+	defaultMaxMetadataValueSize = 128
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/plugin/nodeattestor/gcp/iit.go]: Ugg. Nice catch!

File: pkg/server/plugin/nodeattestor/gcp/iit.go [hunk 2]
```diff
@@ -186,6 +205,54 @@ func (p *IITAttestorPlugin) getConfig() (*IITAttestorConfig, error) {
...
+	for _, item := range instance.Metadata.Items {
+		if !allowedKeys[item.Key] {
+			continue
+		}
+
+		var value string
+		if item.Value != nil {
+			value = *item.Value
+			if len(value) > maxValueSize {
+				p.log.Warn("skipping metadata item %q; value too large (%d > %d)", item.Key, len(value), maxValueSize)
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/server/plugin/nodeattestor/gcp/iit.go]: I agree. Will fix.

**Comment by azdagron** [comment]
I think you are right @APTy. I think I'd rather be restrictive at this point and allow the "valve" to be "opened wide" later, if necessary.

File: doc/plugin_server_nodeattestor_gcp_iit.md [hunk 3]
```diff
@@ -39,6 +42,23 @@ If `use_instance_metadata` is true, then the Google Compute Engine API is querie
 | -------------------------- | ------------------------------------------------------------ | --------------------------------------------------------------------------------|
 | `gcp_iit:tag`              | `gcp_iit:tag:blog-server`                                    | Instance tag (one selector per)
 | `gcp_iit:sa`               | `gcp_iit:sa:123456789-compute@developer.gserviceaccount.com` | Service account (one selector per) 
+| `gcp_iit:label`            | `gcp_iit:label:key:value`                                    | Instance label
+| `gcp_iit:metadata`         | `gcp_iit:metadata:key:value`                                 | Instance metadata (see caveat below)
+
+Not all instance label and metadata values are useful for node selection. To
+prevent the creation of large amounts of useless selectors, labels and metadata
+are not used by default. To opt-in to use a specific label or metadata value,
+specify the key in the `allowed_label_keys` or `allowed_metadata_keys`
```
Comment (APTy) [inline doc/plugin_server_nodeattestor_gcp_iit.md]: Should we note that in order for `allowed_metadata_keys` to work, the `use_instance_metadata` config flag must also be set?

Comment (azdagron) [reply doc/plugin_server_nodeattestor_gcp_iit.md]: It's hard to see in the diff, but the selector table is split into those provided by the IIT and those provided by the instance metadata, separated by the following:

> If `use_instance_metadata` is true, then the Google Compute Engine API is queried for instance metadata which is used to populate these additional selectors:

**Pull request merged**